### PR TITLE
[#16086][fix] Skip value-irrelevant in-place initializers under MetaInitMode

### LINKED ISSUE #16086: [Bug]: 142 GiB/rank host memory leak (leading to host OOM if KV cache offloading is enabled) when serving Kimi-K2.6-NVFP4 TP4 on GB300 (aarch64), due to MetaInitException on aten.fill_.Scalar (nn.LayerNorm init) silently falling back to whole-model CPU construction
## System Info

- Hardware: GB300 (Grace-Blackwell) node, aarch64, 4× GPU, TP4
- Container: `nvcr.io/nvidia/tensorrt-llm/release:1.3.0rc4` (also reproduces on current `main` — the relevant code is unchanged)
- torch: 2.10 NGC build (CPU allocator = mimalloc v2.2.4 on aarch64, enabled by pytorch/pytorch#164741 — note that PR shows as *Closed* because pytorch merges via pytorchmergebot, and it was briefly reverted then relanded the same day 2025-10-09; current pytorch `main` still sets `USE_MIMALLOC ON` for `CPU_AARCH64` in `CMakeLists.txt`, and the NGC build logs `mimalloc: v2.2.4` under `MIMALLOC_VERBOSE=1`)
- Model: [`nvidia/Kimi-K2.6-NVFP4`](https://huggingface.co/nvidia/Kimi-K2.6-NVFP4) (≈554 GB checkpoint; any model whose modules construct stock `nn.LayerNorm` is affected)
- PyTorch backend (`tensorrt_llm._torch`), KV cache host offloading enabled (`KvCacheConfig.host_cache_size`)

## Who can help?

Anyone familiar with the PyTorch-backend model loader / `MetaInitMode` (`tensorrt_llm/_torch/models/modeling_utils.py`).

## Information

- [ ] The official example scripts
- [x] My own modified scripts

## Tasks

- [x] My own task or dataset (give details below)

## Reproduction

Trigger (any platform, CPU-only, a few seconds):

```python
import torch.nn as nn
from tensorrt_llm._torch.models.modeling_utils import MetaInitMode

with MetaInitMode():
    nn.LayerNorm(64)   # raises MetaInitException: Meta tensor used in unsupported function: aten.fill_.Scalar
```

Deployment-scale consequence (aarch64):

1. Serve a large checkpoint whose model constructs `nn.LayerNorm` in `__init__` (e.g. Kimi-K2.6's vision tower) with the PyTorch backend on an aarch64 node.
2. Engine log shows, once per rank per start (at `info` level only): `Fallback to regular model init: ... MetaInitException: Meta tensor used in unsupported function: aten.fill_.Scalar`
3. After weights land on GPU, check `grep RssAnon /proc/<rank_pid>/status` (RssAnon = anonymous resident set size, the process's actual physical-RAM footprint from heap/allocator arenas): each rank retains a weight-shard-sized private-anonymous region for process lifetime (137-142 GiB/rank for the ≈554 GB TP4 checkpoint; ≈568 GiB/node, 54% of host RAM), invisible to allocator statistics.
4. Allocator discriminator proving the pages are freed-but-retained (no rebuild):

```bash
MIMALLOC_VERBOSE=1 MIMALLOC_PURGE_DELAY=0 MIMALLOC_ABANDONED_PAGE_PURGE=1 <launch engine>
# log shows "mimalloc: v2.2.4" (env read); post-init RssAnon collapses to 5-6 GiB/rank
```

## Expected behavior

Meta init survives value-irrelevant in-place initializers (`fill_`/`zero_`) that stock torch modules run in `__init__`; the model constructs on the meta device; no CPU materialization of the weights; host memory after load is a few GiB per rank.

## actual behavior

`MetaInitMode` raises on `aten.fill_.Scalar` / `aten.zero_.default`, the loader's broad `except Exception` silently constructs the ENTIRE model on CPU memory, and on aarch64 the freed CPU copy is retained by mimalloc for process lifetime:

- 142 GiB/rank of dead (freed-but-retained) host memory, ≈568 GiB per node, for the ≈554 GB TP4 checkpoint;
- `KvCacheConfig.host_cache_size` cannot be raised beyond ≈30 GB/rank without host OOM, on a node that should support 150 GB+/rank;
- load time also regresses on every platform (full CPU init instead of meta skeleton).

Root cause of the retention half is upstream of TensorRT-LLM (mimalloc v2.2.4 arena purge gate; same class as the open pytorch/pytorch#178726, proposed fix pytorch/pytorch#178851) — but TensorRT-LLM turns a benign in-place init into that trap by aborting meta init on ops whose results are never consumed.

## additional notes

- A fix PR is filed right after this issue (branch `Weili-0234/TensorRT-LLM:fix-metainit-inplace-init-ops`, will be auto-linked here): add a `deterministic_init_ops` pass-through (`fill_.Scalar`, `fill_.Tensor`, `zero_.default`) to `MetaInitMode`, plus regression tests (`tests/unittest/_torch/models/test_meta_init_mode.py`). Red/green verified on a GB300 node with the rc4 image: 3 failed + 1 passed before, 4 passed after.
- Searchable log signatures for others hitting this: `Fallback to regular model init`, `MetaInitException: Meta tensor used in unsupported function: aten.fill_.Scalar`, unexplained per-rank physical memory (`RssAnon`) ≈ weight-shard size on Grace (GB200/GB300).
- Suggested follow-up (not in the PR): log the whole-model CPU fallback at `warning` instead of `info` — at production log levels the fallback is currently invisible, which is why this took a full allocation-stack investigation to find.

### PR DESCRIPTION
**Pull request by Weili-0234**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved meta-tensor initialization so common deterministic in-place setup operations no longer fail during model construction.
  * Preserved expected aliasing behavior for in-place initialization on meta tensors.
  * Added coverage to confirm models using zero-initialized parameters can be built on the meta device and later materialized correctly.
  * Continued to surface errors for unsupported meta-tensor operations, so fallback behavior still works as expected.

Closes #16086

### Symptom

Deploying [`nvidia/Kimi-K2.6-NVFP4`](https://huggingface.co/nvidia/Kimi-K2.6-NVFP4) (≈554 GB, TP4) on a GB300 (Grace-Blackwell, aarch64) node with KV cache host offloading, we could not raise `KvCacheConfig.host_cache_size` beyond ≈30 GB per rank without host OOM — on a node whose RAM should comfortably support 150 GB+ per rank. Each rank process carried a **142 GiB private-anonymous region** (≈568 GiB/node, 54% of host RAM) that was invisible to every allocator statistic, present from model init onward.

### Root cause

1. Stock torch modules run value-irrelevant in-place initializers in `__init__`: `nn.LayerNorm.reset_parameters` → `init.ones_` → `aten.fill_.Scalar` (Kimi-K2.6's vision tower constructs `nn.LayerNorm` directly). `MetaInitMode.__torch_dispatch__` raises `MetaInitException` on these ops.

2. The broad `except Exception` around meta init in `model_loader.py` then silently falls back to constructing the **entire model on CPU memory** (logged at `info` level only):

   ```
   Fallback to regular model init: ...
     torch/nn/modules/normalization.py reset_parameters -> init.ones_
   MetaInitException: Meta tensor used in unsupported function: aten.fill_.Scalar
   ```

3. On aarch64 the CPU copy becomes a permanent leak: NGC torch ≥ 2.10 builds use mimalloc as the default CPU allocator on aarch64 (pytorch/pytorch#164741). The CPU tensors are freed during `model.to('cuda')`, but mimalloc v2.2.4 retains the dirty weight-shard-sized arena pages for process lifetime (arena purge gate defect in `mi_arenas_try_purge`; same class as the open pytorch/pytorch#178726 "5× RSS regression on aarch64", proposed fix pytorch/pytorch#178851). Empirical discriminator, no rebuild needed:

   ```bash
   MIMALLOC_VERBOSE=1 MIMALLOC_PURGE_DELAY=0 MIMALLOC_ABANDONED_PAGE_PURGE=1 <launch engine>
   # log shows "mimalloc: v2.2.4" (env read); post-init RssAnon per rank
   # collapses from ~137.6 GiB to ~5-6 GiB — proving freed-but-retained pages.
   ```

On x86-64 (glibc) the fallback costs load time and transient peak RSS only, so the bug hides there; aarch64 turns it into permanent host-RAM loss.

Minimal repro of the trigger (any platform, CPU-only):

```python
import torch.nn as nn
from tensorrt_llm._torch.models.modeling_utils import MetaInitMode

with MetaInitMode():
    nn.LayerNorm(64)   # raises MetaInitException before this fix
```

### Fix

Add a `deterministic_init_ops` set (`aten.fill_.Scalar`, `aten.fill_.Tensor`, `aten.zero_.default`) to `MetaInitMode` and skip these ops on meta tensors, returning the target tensor to keep in-place aliasing semantics. Their results are placeholders that are always overwritten after construction (dummy init or checkpoint load), so skipping is value-safe. Ops whose results are consumed (e.g. computing non-persistent buffers) still raise `MetaInitException`, preserving the conservative fallback.

### Verification

Red/green on a GB300 node (aarch64, 4× GPU), image `nvcr.io/nvidia/tensorrt-llm/release:1.3.0rc4`:

- **Red** (installed package, verbatim): `test_layernorm_constructs_on_meta`, `test_zeros_init_constructs_on_meta`, `test_meta_params_materialize_after_init` fail with `MetaInitException`; `test_unsupported_op_still_raises` passes.
- **Green** (this patch applied): all 4 pass.
- Normal-path sanity: `nn.LayerNorm` outside `MetaInitMode` still initializes to ones on CPU; `fill_` on real CUDA tensors unaffected.
- At deployment scale (the same Kimi-K2.6-NVFP4 TP4 engine, aarch64): zero fallback lines in the engine log; post-init `RssAnon` per rank drops from ≈137.6 GiB to 7.8-11.2 GiB; the same node could then run `host_cache_size` 5× larger.

## Test Coverage

New `tests/unittest/_torch/models/test_meta_init_mode.py` (CPU-only, no GPU required):

- `test_layernorm_constructs_on_meta` — the `fill_.Scalar` trigger; fails on any tree without the pass-through.
- `test_zeros_init_constructs_on_meta` — the `zero_.default` trigger.
- `test_meta_params_materialize_after_init` — meta params still materialize via `_apply`.
- `test_unsupported_op_still_raises` — the conservative contract is pinned: consumed-result ops (`torch.mm`) still raise.

```bash
python3 -m pytest tests/unittest/_torch/models/test_meta_init_mode.py -v
```

## PR Checklist

- [x] PR description clearly explains what and why.
- [x] PR follows TRT-LLM coding guidelines (pre-commit hooks pass on both files; ruff/ruff-format applied to the new test; legacy-file rules respected for `modeling_utils.py`).
- [x] Test cases are provided for new code paths (red/green pair above).
- [x] No new dependencies.
- [x] Commit is DCO signed-off.

### CODE REVIEW

**Comment by Weili-0234** [comment]
hi @yechank-nvidia, could you please take a look and trigger the CI if possible?

File: tensorrt_llm/_torch/models/modeling_utils.py [hunk 1]
```diff
@@ -90,6 +108,12 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
                 kwargs = {}
             kwargs['device'] = torch.device('meta')
             return func(*args, **kwargs)
+        elif func in self.deterministic_init_ops and self._has_meta_tensor(
+                args, kwargs):
+            # In-place init on a meta tensor: the value is irrelevant (it is
+            # overwritten after construction), so skip the op and return the
+            # target tensor to keep in-place aliasing semantics.
+            return args[0]
         elif func not in self.random_init_ops and self._has_meta_tensor(
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_utils.py]: Suggestion: instead of skipping and returning args[0], exempt these ops from the raise guard so they fall through and run on the meta kernel — same pattern as random_init_ops,

```suggestion
        elif func not in self.random_init_ops and \
                  func not in self.deterministic_init_ops and \
                  self._has_meta_tensor(args, kwargs):
              raise MetaInitException(
                  f"Meta tensor used in unsupported function: {func}")
```

Worst case for a bad future addition is a loud `NotImplementedError` in the meta-init tests, not silently wrong values.

File: tensorrt_llm/_torch/models/modeling_utils.py [hunk 2]
```diff
@@ -75,6 +75,24 @@ class MetaInitMode(TorchDispatchMode):
...
+    # TP4 on GB300). Their results are placeholders that are overwritten
+    # after construction (dummy init or checkpoint load), so they are safe
+    # to skip on meta tensors. Do NOT add ops whose results are consumed
+    # (e.g. ops computing non-persistent buffers): the conservative
+    # MetaInitException fallback below still protects those.
+    deterministic_init_ops = {
+        aten.fill_.Scalar,
+        aten.fill_.Tensor,
+        aten.zero_.default,
+    }
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_utils.py]: Check if addressed

File: tests/unittest/_torch/models/test_meta_init_mode.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def test_unsupported_op_still_raises():
+    # The conservative contract is load-bearing: ops whose results are
+    # actually consumed (e.g. computing non-persistent buffers) must still
+    # abort meta init so the caller can fall back to regular construction.
+    with MetaInitMode():
+        t = torch.empty(4, 4)  # redirected to the meta device
+        with pytest.raises(MetaInitException):
+            torch.mm(t, t)
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tests/unittest/_torch/models/test_meta_init_mode.py]: Check if addressed

File: tests/unittest/_torch/models/test_meta_init_mode.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+def test_zeros_init_constructs_on_meta():
+    class ZeroBias(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.bias = nn.Parameter(torch.empty(16))
+            nn.init.zeros_(self.bias)
+
+    with MetaInitMode():
+        m = ZeroBias()
+    assert m.bias.device == torch.device("meta")
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tests/unittest/_torch/models/test_meta_init_mode.py]: Check if addressed

**Review by yechank-nvidia — APPROVED** [review]
